File: server-legacy/src/main/java/de/pnp/manager/model/Fabrication.java
package de.pnp.manager.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import de.pnp.manager.main.Database;
import de.pnp.manager.model.item.IItem;
import de.pnp.manager.model.item.Item;

import java.util.Objects;
import java.util.stream.Collectors;

public class Fabrication implements IFabrication {

    protected IItem product = Item.EMPTY_ITEM;
    protected String profession = "";
    protected String requirement = "";
    protected String otherCircumstances = "";
    private float productAmount = 1;
    private IItem sideProduct = Item.EMPTY_ITEM;
    private float sideProductAmount = 1;
    private IItemList materials = new ItemList();

    public IItem getProduct() {
        return product;
    }

    public void setProduct(IItem product) {
        this.product = product;
    }

    public String getProfession() {
        return profession;
    }

    public void setProfession(String profession) {
        this.profession = profession;
    }

    public String getRequirement() {
        return requirement;
    }

    public void setRequirement(String requirement) {
        this.requirement = requirement;
    }

    public String getOtherCircumstances() {
        return otherCircumstances;
    }

    public void setOtherCircumstances(String otherCircumstances) {
        this.otherCircumstances = otherCircumstances;
    }

    public float getProductAmount() {
        return productAmount;
    }

    public void setProductAmount(float productAmount) {
        this.productAmount = productAmount;
    }

    public IItem getSideProduct() {
        return sideProduct;
    }

    public void setSideProduct(IItem sideProduct) {
        this.sideProduct = sideProduct;
    }

    public float getSideProductAmount() {
        return sideProductAmount;
    }

    public void setSideProductAmount(float sideProductAmount) {
        this.sideProductAmount = sideProductAmount;
    }

    @JsonIgnore
    public String getProductName() {
        return Objects.requireNonNullElse(product.getName(), "");
    }

    public void setProductName(String itemName) {
        product = Database.getItem(itemName);
    }

    @JsonIgnore
    public String getSideProductName() {
        return Objects.requireNonNullElse(sideProduct.getName(), "");
    }

    public void setSideProductName(String itemName) {
        sideProduct = Database.getItem(itemName);
    }

    public IItemList getMaterials() {
        return materials;
    }

    public void setMaterials(IItemList materials) {
        this.materials = materials;
    }

    @JsonIgnore
    public String getMaterialsAsString() {
        return materials.stream().map(item -> item.getPrettyAmount() + " " + item.getName()).collect(Collectors.joining("\n"));
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof Fabrication)) {
            return false;
        }

        Fabrication other = (Fabrication) obj;

        return getProduct().equals(other.getProduct()) && getProfession().equals(other.getProfession())
                && getRequirement().equals(other.getRequirement()) && getOtherCircumstances().equals(other.getOtherCircumstances())
                && getProductAmount() == other.getProductAmount() && getSideProduct().equals(other.getSideProduct())
                && getSideProductAmount() == other.getSideProductAmount() && getMaterials().equals(other.getMaterials());
    }
}

File: app/src/main/java/de/pnp/manager/app/model/Fabrication.kt
package de.pnp.manager.app.model

import de.pnp.manager.model.IFabrication
import de.pnp.manager.model.IItemList
import de.pnp.manager.model.item.IItem

class Fabrication : IFabrication {
    private var product : IItem? = null
    private var profession : String = ""
    private var requirement: String = ""
    private var otherCircumstances : String = ""
    private var productAmount: Float = 0f
    private var sideProduct : IItem? = null
    private var sideProductAmount: Float = 0f
    private var materials: IItemList? = null

    override fun getProduct(): IItem? {
        return this.product
    }

    override fun setProduct(product: IItem?) {
        this.product = product
    }

    override fun getProfession(): String {
        return this.profession
    }

    override fun setProfession(profession: String) {
        this.profession = profession
    }

    override fun getRequirement(): String {
        return requirement
    }

    override fun setRequirement(requirement: String) {
        this.requirement = requirement
    }

    override fun getOtherCircumstances(): String {
        return otherCircumstances
    }

    override fun setOtherCircumstances(otherCircumstances: String) {
        this.otherCircumstances = otherCircumstances
    }

    override fun getProductAmount(): Float {
        return productAmount
    }

    override fun setProductAmount(productAmount: Float) {
        this.productAmount = productAmount
    }

    override fun getSideProduct(): IItem? {
        return sideProduct
    }

    override fun setSideProduct(sideProduct: IItem?) {
        this.sideProduct = sideProduct
    }

    override fun getSideProductAmount(): Float {
        return this.sideProductAmount
    }

    override fun setSideProductAmount(sideProductAmount: Float) {
        this.sideProductAmount = sideProductAmount
    }

    override fun getMaterials(): IItemList? {
        return materials
    }

    override fun setMaterials(materials: IItemList?) {
        this.materials = materials
    }
}